akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/LoggingEventFilterImpl.scala

```diff
@@ -0,0 +1,166 @@
+ */
+
+package akka.actor.testkit.typed.internal
+
+import java.util.function.Supplier
+
+import scala.concurrent.duration.Duration
+import scala.reflect.ClassTag
+import scala.util.matching.Regex
+
+import akka.actor.testkit.typed.LoggingEvent
+import akka.actor.testkit.typed.TestKitSettings
+import akka.actor.testkit.typed.javadsl
+import akka.actor.testkit.typed.scaladsl
+import akka.actor.typed.ActorSystem
+import akka.annotation.InternalApi
+import akka.testkit.TestKit
+import org.slf4j.event.Level
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object LoggingEventFilterImpl {
+  def empty: LoggingEventFilterImpl = new LoggingEventFilterImpl(1, None, None, None, None, None, Map.empty, None)
+}
+
+/**
+ * INTERNAL API
+ *
+ * Facilities for selectively filtering out expected events from logging so
+ * that you can keep your test run’s console output clean and do not miss real
+ * error messages.
+ *
+ * See the companion object for convenient factory methods.
+ *
+ * If the `occurrences` is set to Int.MaxValue, no tracking is done.
+ */
+@InternalApi private[akka] final case class LoggingEventFilterImpl(
+    occurrences: Int,
+    logLevel: Option[Level],
+    source: Option[String],
+    messageContains: Option[String],
+    messageRegex: Option[Regex],
+    cause: Option[Class[_ <: Throwable]],
+    mdc: Map[String, String],
+    custom: Option[Function[LoggingEvent, Boolean]])
+    extends javadsl.LoggingEventFilter
+    with scaladsl.LoggingEventFilter {
+
+  @volatile // JMM does not guarantee visibility for non-final fields
+  private var todo = occurrences
+
+  /**
+   * This method decides whether to filter the event (<code>true</code>) or not
+   * (<code>false</code>).
+   */
+  def matches(event: LoggingEvent): Boolean = {
+    logLevel.forall(_ == event.level) &&
+    source.forall(_ == sourceOrEmpty(event)) &&
+    messageContains.forall(messageOrEmpty(event).contains) &&
+    messageRegex.forall(_.findFirstIn(messageOrEmpty(event)).isDefined) &&
+    cause.forall(c => event.throwable.isDefined && c.isInstance(event.throwable.get)) &&
+    mdc.forall { case (key, value) => event.mdc.contains(key) && event.mdc(key) == value } &&
+    custom.forall(f => f(event))
+  }
+
+  private def messageOrEmpty(event: LoggingEvent): String =
+    if (event.message == null) "" else event.message
+
+  private def sourceOrEmpty(event: LoggingEvent): String =
+    event.mdc.getOrElse("akkaSource", "")
+
+  def apply(event: LoggingEvent): Boolean = {
+    if (matches(event)) {
+      if (todo != Int.MaxValue) todo -= 1
+      true
+    } else false
+  }
+
+  def awaitDone(max: Duration): Boolean = {
+    if (todo != Int.MaxValue && todo > 0) TestKit.awaitCond(todo <= 0, max, noThrow = true)
+    todo == Int.MaxValue || todo == 0
+  }
+
+  /**
+   * Assert that this filter has matched as often as requested by its
+   * `occurrences` parameter specifies.
+   */
+  def assertDone(max: Duration): Unit =
```

Comment:
Just internal or missing an `override`?

Reply:
entire LoggingEventFilterImpl is internal

Reply:
but it wasn't used, so removed

Reply:
Was a bit confusing since this very much looks like it was one of the public methods, perhaps just `private` it then?

Reply:
removed it, and made one of the other private